cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -247,4 +269,34 @@ public WorkflowToken deepCopy() {
     }
     return builder.build();
   }
+
+  private void ensureEnoughMemoryAvailable(String key, Value value) {
+    int left = bytesLeft;
+    left -= key.length();
+    if (left <= 0) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB' while " +
+                                                      "adding key '%s'. Current size is '%sMB'. Please increase the " +
+                                                      "maximum permitted size by setting the parameter '%s' in " +
+                                                      "cdap-site.xml to add more values.",
+                                                    maxSizeBytes / (1024 * 1024), key,
+                                                    currentSizeBytes / (1024 * 1024),
+                                                    Constants.AppFabric.WORKFLOW_TOKEN_MAX_SIZE_MB));
+    }
+    left -= value.toString().length();
+    if (left <= 0) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB' while " +
+                                                      "adding key '%s' with value '%s'. Current size is '%sMB'. " +
+                                                      "Please increase the maximum permitted size by setting the " +
+                                                      "parameter '%s' in cdap-site.xml to add more values.",
+                                                    maxSizeBytes / (1024 * 1024), key, value,
+                                                    currentSizeBytes / (1024 * 1024),
+                                                    Constants.AppFabric.WORKFLOW_TOKEN_MAX_SIZE_MB));
+    }
+  }
+
+  private void updateSize(String key, Value value) {
```

Comment:
Why need to separate as two? Can it simply just do compute the potential new size and do the update+put? it's just 1 line each for the update and put.